File: Server-CSharp/Server-CSharp/Demo/Handler/MainMsgHandle.cs
using Game;
using Google.Protobuf;
using System;
using YT;

namespace YT
{
    public partial class MsgHandler
    {

        //获取个人简介内容
        public static void MsgGetPlayerIntroduction(Connection c, byte[] bytes) {
            MsgGetPlayerIntroduction msg = Game.MsgGetPlayerIntroduction.Parser.ParseFrom(bytes);
            Player player = c.player;
            if (player == null) return;
            //获取text
            msg.PalyerIntroduction = player.data.palyerIntroduction;
            player.Send(msg);
        }

        //保存个人简介内容
        public static void MsgSavePlayerIntroduction(Connection c, byte[] bytes) {
            MsgSavePlayerIntroduction msg = Game.MsgSavePlayerIntroduction.Parser.ParseFrom(bytes);
            Player player = c.player;
            if (player == null) return;
            //获取头像
            player.data.palyerIntroduction = msg.PalyerIntroduction;
            //保存数据
            DBManager.UpdatePlayerData(c.player.id, c.player.data);
            player.Send(msg);
        }

        //获取头像
        public static void MsgGetHeadPhoto(Connection c, byte[] bytes) {
            MsgGetHeadPhoto msg = Game.MsgGetHeadPhoto.Parser.ParseFrom(bytes);
            Player player = c.player;
            if (player == null) return;
            //获取头像
            msg.HeadPhoto = player.data.headPhoto;
            player.Send(msg);
        }

        //保存头像
        public static void MsgSaveHeadPhoto(Connection c, byte[] bytes) {
            MsgSaveHeadPhoto msg = Game.MsgSaveHeadPhoto.Parser.ParseFrom(bytes);
            Player player = c.player;
            if (player == null) return;
            //获取头像
            player.data.headPhoto = msg.HeadPhoto;
            //保存数据
            DBManager.UpdatePlayerData(c.player.id, c.player.data);
            player.Send(msg);
        }

        //发送消息到世界
        public static void MsgSendMessageToWord(Connection c, byte[] bytes) {
            MsgSendMessageToWord msg = Game.MsgSendMessageToWord.Parser.ParseFrom(bytes);

            Player player = c.player;
            //返回协议
            msg.Result = 0;
            PlayerManager.Broadcast(msg);


        }

        ////发送消息到好友
        public static void MsgSendMessageToFriend(Connection c, byte[] bytes) {
            MsgSendMessageToFriend msg = Game.MsgSendMessageToFriend.Parser.ParseFrom(bytes);


        }

        //获取好友列表
        public static void MsgGetFriendList(Connection c, byte[] bytes) {
            MsgGetFriendList msg = Game.MsgGetFriendList.Parser.ParseFrom(bytes);
            Player player = c.player;
            if (player == null) return;
            //获取FriendList
            msg.FriendList = DBManager.GetFriendList(c.player.id);
            msg.Result = 0;
            player.Send(msg);
        }
        //添加好友
        public static void MsgAddFriend(Connection c, byte[] bytes) {
            MsgAddFriend msg = Game.MsgAddFriend.Parser.ParseFrom(bytes);
            Player player = c.player;
            if (player == null) return;

            if (DBManager.IsAccountExist(msg.FriendId)) {
                msg.Result = 1;
                player.Send(msg);
                return;
            }
            if (!DBManager.IsFriendExist(msg.Id, msg.FriendId)) {
                msg.Result = 3;
                player.Send(msg);
                return;
            }
            if (PlayerManager.GetPlayer(msg.FriendId) == null) {
                msg.Result = 2;
                player.Send(msg);
                return;
            }
            msg.Result = 0;
            player.Send(msg);
            PlayerManager.GetPlayer(msg.FriendId).Send(msg);
        }

        //删除好友
        public static void MsgDeleteFriend(Connection c, byte[] bytes) {
            MsgDeleteFriend msg = Game.MsgDeleteFriend.Parser.ParseFrom(bytes);
            Player player = c.player;
            if (player == null) return;
            if (DBManager.DeleteFriend(player.id, msg.FriendId)) {
                msg.Result = 0;
                player.Send(msg);
                if (PlayerManager.GetPlayer(msg.FriendId) != null) {
                    PlayerManager.GetPlayer(msg.FriendId).Send(msg);
                }

            }
            else {
                msg.Result = 1;
                player.Send(msg);
            }


        }
        //同意添加好友
        public static void MsgAcceptAddFriend(Connection c, byte[] bytes) {
            MsgAcceptAddFriend msg = Game.MsgAcceptAddFriend.Parser.ParseFrom(bytes);
            Player player = c.player;
            if (player == null) return;

            if (DBManager.AddFriend(msg.Id, msg.FriendId)) {
                msg.Result = 0;
            }
            else {
                msg.Result = 1;
            }
            player.Send(msg);
            PlayerManager.GetPlayer(msg.FriendId).Send(msg);
        }
    }
}

File: Server-Go/Demo/Handler/MainMsgHandle.go
package Handler

import (
	"ReaminHistory/Demo/Player/PlayerManager"
	"ReaminHistory/Demo/Proto"
	"ReaminHistory/Demo/db"
	"ReaminHistory/YT"
	"fmt"
	"github.com/golang/protobuf/proto"
)

//获取个人简介内容
func MsgGetPlayerIntroduction(c *YT.Connection, bytes []byte) {
	msg := &Proto.MsgGetPlayerIntroduction{}
	err := proto.Unmarshal(bytes, msg)
	if err != nil {
		fmt.Println("解析失败")
		return
	}
	if player, ok := PlayerManager.Players[c.Conv]; ok {
		msg.PalyerIntroduction = player.Data.PlayerIntroduction
		player.Send(msg)
	} else {
		return
	}
}

//保存个人简介内容
func MsgSavePlayerIntroduction(c *YT.Connection, bytes []byte) {
	msg := &Proto.MsgSavePlayerIntroduction{}
	err := proto.Unmarshal(bytes, msg)
	if err != nil {
		fmt.Println("解析失败")
		return
	}
	if player, ok := PlayerManager.Players[c.Conv]; ok {
		player.Data.PlayerIntroduction = msg.PalyerIntroduction
		db.UpdatePlayerData(player.Id, player.Data)
		player.Send(msg)
	} else {
		return
	}
}

//获取头像
func MsgGetHeadPhoto(c *YT.Connection, bytes []byte) {
	msg := &Proto.MsgGetHeadPhoto{}
	err := proto.Unmarshal(bytes, msg)
	if err != nil {
		fmt.Println("解析失败")
		return
	}
	if player, ok := PlayerManager.Players[c.Conv]; ok {
		msg.HeadPhoto = player.Data.HeadPhoto
		player.Send(msg)
	} else {
		return
	}
}

//保存头像
func MsgSaveHeadPhoto(c *YT.Connection, bytes []byte) {
	msg := &Proto.MsgSaveHeadPhoto{}
	err := proto.Unmarshal(bytes, msg)
	if err != nil {
		fmt.Println("解析失败")
		return
	}
	if player, ok := PlayerManager.Players[c.Conv]; ok {
		player.Data.HeadPhoto = msg.HeadPhoto
		db.UpdatePlayerData(player.Id, player.Data)
		player.Send(msg)
	} else {
		return
	}
}

//发送消息到世界
func MsgSendMessageToWord(c *YT.Connection, bytes []byte) {
	msg := &Proto.MsgSendMessageToWord{}
	err := proto.Unmarshal(bytes, msg)
	if err != nil {
		fmt.Println("解析失败")
		return
	}
	if _, ok := PlayerManager.Players[c.Conv]; ok {
		msg.Result = 0
		PlayerManager.Broadcast(msg)
	} else {
		return
	}
}

//发送消息到好友
func MsgSendMessageToFriend(c *YT.Connection, bytes []byte) {
	msg := &Proto.MsgSendMessageToFriend{}
	err := proto.Unmarshal(bytes, msg)
	if err != nil {
		fmt.Println("解析失败")
		return
	}
	//TODO
}

//获取好友列表
func MsgGetFriendList(c *YT.Connection, bytes []byte) {
	msg := &Proto.MsgGetFriendList{}
	err := proto.Unmarshal(bytes, msg)
	if err != nil {
		fmt.Println("解析失败")
		return
	}
	if player, ok := PlayerManager.Players[c.Conv]; ok {
		msg.FriendList = db.GetFriendList(player.Id)
		msg.Result = 0
		player.Send(msg)
	} else {
		return
	}
}

//添加好友
func MsgAddFriend(c *YT.Connection, bytes []byte) {
	msg := &Proto.MsgAddFriend{}
	err := proto.Unmarshal(bytes, msg)
	if err != nil {
		fmt.Println("解析失败")
		return
	}
	if player, ok := PlayerManager.Players[c.Conv]; ok {
		if db.IsAccountExist(msg.FriendId) {
			msg.Result = 1
			player.Send(msg)
			return
		}
		if db.IsFriendExist(msg.Id, msg.FriendId) {
			msg.Result = 2
			player.Send(msg)
			return
		}
		if PlayerManager.GetPlayer(msg.FriendId) == nil {
			msg.Result = 3
			player.Send(msg)
			return
		}
		msg.Result = 0
		player.Send(msg)
		PlayerManager.GetPlayer(msg.FriendId).Send(msg)
	} else {
		return
	}
}

//删除好友
func MsgDeleteFriend(c *YT.Connection, bytes []byte) {
	msg := &Proto.MsgDeleteFriend{}
	err := proto.Unmarshal(bytes, msg)
	if err != nil {
		fmt.Println("解析失败")
		return
	}
	if player, ok := PlayerManager.Players[c.Conv]; ok {
		if db.DeleteFriend(msg.Id, msg.FriendId) {
			msg.Result = 0
			player.Send(msg)
			if PlayerManager.GetPlayer(msg.FriendId) != nil {
				PlayerManager.GetPlayer(msg.FriendId).Send(msg)
			}
		} else {
			msg.Result = 1
			player.Send(msg)
		}
	} else {
		return
	}
}

//同意添加好友
func MsgAcceptAddFriend(c *YT.Connection, bytes []byte) {
	msg := &Proto.MsgAcceptAddFriend{}
	err := proto.Unmarshal(bytes, msg)
	if err != nil {
		fmt.Println("解析失败")
		return
	}
	if player, ok := PlayerManager.Players[c.Conv]; ok {
		if db.AddFriend(msg.Id, msg.FriendId) {
			msg.Result = 0
		} else {
			msg.Result = 1
		}
		player.Send(msg)
		PlayerManager.GetPlayer(msg.FriendId).Send(msg)
	} else {
		return
	}
}
